src/hashing.rs

```diff
@@ -427,6 +461,48 @@ pub trait DerivedKeyDomain: DomainSeparation {
     }
 }
 
+/// Creates a DomainSeparation struct for a given domain.
+#[macro_export]
+macro_rules! hash_domain {
+    ($name:ident, $domain:expr, $version: expr) => {
+        pub struct $name {}
+
+        impl $crate::hashing::DomainSeparation for $name {
+            fn version() -> u8 {
+                $version
+            }
+
+            fn domain() -> &'static str {
+                $domain
+            }
+        }
+    };
+    ($name:ident, $domain:expr) => {
+        hash_domain!($name, $domain, 1);
+    };
+}
+
+/// Creates a domain separated hasher.
+#[macro_export]
+macro_rules! create_hasher {
+    ($digest:ty, $name:ident, $domain:expr, $label:expr, $version: expr) => {{
+        use $crate::hash_domain;
+
+        hash_domain!($name, $domain, $version);
```

Comment:
My only concern about `create_hasher!` is that users will duplicate the domain separation domain label.

```rust
fn hash_foo(foo:&Foo) -> Hash {
   let hasher = create_hasher!(Blake256, FooBarHasher, "com.foobar", "bar");
   hasher.chain(b"blah").finalize()
}

fn hash_bar(bar:&Bar) -> Hash {
   let hasher = create_hasher!(Blake256, FooBarHasher, "com.foobar", "bar");
   hasher.chain(b"blah").finalize()
}
```

Perhaps `create_hasher!` should take in a domain type?

```rust
define_hash_domain!(FooBarHashDomain, "com.foobar");

fn hash_foo(foo:&Foo) -> Hash {
   let hasher = create_hasher!(Blake256, FooBarHashDomain, "bar");
   hasher.chain(b"blah").finalize()
}

fn hash_bar(bar:&Bar) -> Hash {
   let hasher = create_hasher!(Blake256, FooBarHashDomain, "bar");
   hasher.chain(b"blah").finalize()
}
```

Reply:
The macro isn't really needed unless you are creating a new struct. (The DomainSeparation).

In your second example you can just create a method 

```
fn create_hasher<D:Digest, Dom:DomainSeparation>(label: &'static str) -> ...
```

Reply:
Here's an example of `create_hasher!`, but I agree it's probably better to leave it out and create the method.

```rust 
     create_hasher!(Blake2b, "com.tari.cipher_seed_mac")
            .chain(plaintext.clone())
            .chain([CIPHER_SEED_VERSION])
            .chain(self.salt)
            .chain(passphrase.as_bytes())
            .finalize()
```

Reply:
same concern as @sdbondi, it's nice to add a suggestive comment not the use the long form macro `create_hasher!()`, maybe add another macro next to it, that would differ by name instead of only arguments

Reply:
users might not even notice that calling the long form in multiple places will define nested structs in different places where it's called